packages/opentelemetry-core/src/trace/instrumentation/BasePlugin.ts

```diff
@@ -0,0 +1,40 @@
+/**
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Tracer, Plugin } from '@opentelemetry/types';
+
+/** This class represent the base to patch plugin. */
+export abstract class BasePlugin<T> implements Plugin<T> {
```

Comment:
I asked this question on anther PR, but asking it here again for completeness: does it make sense for `BasePlugin` to be part of `opentelemetry-core` given that it's a Node platform specific concept?

Reply:
I am just afraid about circular dependency if we move `BasePlugin` in `Node-Tracer` package because any instrumented plugin (ex: `package/plugin-http`) needs to `extends BasePlugin` from Node-Tracer and add same package in Node-Tracer for enabling the patching.

Reply:
Ah, makes sense! Yeah, I think it's fine then.

Reply:
If it's Node specific, then IMO it definitely shouldn't be in core. Maybe we need a new package to store this? Otherwise, any way we can make sure that the base class is not Node specific?

Reply:
Well, the whole concept of a Plugin as a runtime thing is somewhat Node specific. A couple options:
- Have a package like `node-core` that contains Node-specific shared functionality. There could also be a `web-core` package that has web specific utilities in the future.
- Just make `Plugin` an interface and move it to the `types` package - we lose the little bit of reusable code that it provides, but this isn't hard to duplicate for each plugin.

Reply:
Why would the concept of Plugin not exist in the browser? I would expect to have for example an Angular plugin and a React plugin that potentially hook into any exposed globals. It would work completely differently than in Node, but I think the concept of a plugin itself still applies. Did you have another approach in mind for the browser?

Reply:
Yes, that's a good point - at a high level the browser has the concept of plugins. But I guess I'm thinking of a webpack build (possibly with Closure compiler in the mix). In that case, the plugin functions could be renamed or inlined, etc. by the build process. So the tracer can't really just come in and hook into it at runtime the same way.

I think for the browser the main hooks will be of the core web APIs like `XMLHTTPRequset`, etc. The way I would imagine dealing with that is to hook into those core web APIs before any wrapper libraries get loaded.

For users that want it, [Zone.js](https://github.com/angular/zone.js/) is a nice wrapper for wholesale monkey patching the core browser async APIs.

I'll have to think about this more.

Reply:
My only worry with a compile-time approach is that there are a lot of different bundlers out there, so we can't assume everyone uses Webpack. There are also a lot of apps and libraries out there that do not require compilation at all.

Reply:
Yes, I agree. The challenge with the runtime approach for patching something like a framework is that the build process will have renamed properties in the framework and/or inlined functions (or even whole classes!). So you can't really know how to monkey-patch a framework at runtime.

Our current approach in OpenCensus Web is to only patch the core web APIs, which always have defined names at runtime, and as a plus side will work no matter how you compile your JS and no matter what web framework you use :)

Reply:
But to come back to why I wonder if the `Plugin` class as defined may fit well for the web, those core browser APIs don't have any notion of `require` or modules, they are just globals. (There is some module work in Chrome and new browsers for some new things, but most are still globals)

We could still have Plugin represent the patch/unpatch methods but they would be patching globals not a module object.

Reply:
> the build process will have renamed properties in the framework 

True, but many frameworks expose some sort of global object that is not renamed (otherwise it couldn't be interacted with). I think maybe a good compromise would be having compile-time requirements that will allow the plugins to work at runtime. Then it would be possible to handle the requirements automatically with a Webpack loader, but if the user is not using Webpack they can still do it manually. So basically at compile time, anything needed by the plugin would simply get exposed, and then available at runtime for the plugin to do its magic. That approach would make it a lot easier for users to do the compile-time step manually if needed.

> We could still have Plugin represent the patch/unpatch methods but they would be patching globals not a module object.

I think that's a fair assumption.

Reply:
Those are cool ideas! Definitely worth exploring more once we get deeper into the browser side of this.